### ISSUE TITLE
Using Knet with StaticArrays

### ISSUE THREAD

**Issue by denglerchr** [denizyuret/Knet.jl#305]
It would be a nice feature to enable the speedup of StaticArrays for Knet if it's not too hard to bring this in. It could result in a nice speedup for training if matrices are not too big, it certainly does for inference.
At the moment, the creation of the optimizer returns a Void object, which leads to errors later when calling "optimize!".
```julia
using Knet, StaticArrays

w1 = [randn(5, 5), randn(5)]
println(optimizers(w1, Adam)) # This works

w2 = [SMatrix{5, 5}(randn(5, 5)), SVector{5}(randn(5))]
println(optimizers(w2, Adam)) # This does not produce an error but returns Void
```

**Comment by CarloLucibello** [comment]
I fixed the `optimizers` call  in #306. While it gets merged, you could just define the optimizers like this:
`opt= [Adam() for i=1:length(w)]`
and see how far you can go in your experiment with StaticArrays. 

With small arrays the overhead of AutoGrad can be huge, so I'm not sure there is anything to gain with StaticArrays, nonetheless it is worth to try. 

Also, knet and autograd should be updated to work with any abstract array, another reason why is good to try out StaticArrays

**Comment by CarloLucibello** [comment]
I'm trying to address the problems in update!, but I keep hitting:
```julia
ERROR: setindex!(::StaticArrays.SArray{Tuple{5},Float64,1,5}, value, ::Int) is not defined.
```
Looks like static arrays don't support setindex! (in fact they have the non mutating version `setindex`)

**Comment by denglerchr** [comment]
You are right, I was looking into their package, and it seems that StaticArrays are immutable (data is stored as a tuple), even though in the description they have "Note that here "statically sized" means that the size can be determined from the type, and "static" does not necessarily imply immutable." :( 
So I guess StaticArrays are not an option for regression parameters after all.

**Comment by denglerchr** [comment]
Oh I should have read further: "Mutable versions MVector, MMatrix and MArray are also exported, as well as SizedArray for annotating standard Arrays with static size information." . But according to their benchmarks it doesn't improve much on matrix multiplications, so I guess it wouldn't improve the learning speed by much.
